### ISSUE TITLE
Support constant names for extended enums

### ISSUE THREAD

**Issue by heaths** [Azure/autorest.csharp#855]
Based on archboard feedback during review of Search's FieldBuilder, we discussed support for constant names for extended enums, which use a nested static class with public constants (similar to some track 1 instances). It would be great to have generator support for this, though in the meantime we can write tests using reflection to make sure that extended enum static properties and constant names match exactly.

For should follow:

```csharp
public readonly partial struct ExEnum
{
  private const string FooValue = "foo";
  public static ExEnum Foo  { get; } = new ExEnum(FooValue);
  public static class Names // Names here is fixed; if collision exists,
                            // people can use something like NamesAsString,
                            // but we suspect the chances of this are very small
                            // since this feature should only be used when constant expressions are needed.
  {
    public const string Foo = ExEnum.FooValue;
  }
}
```

The XML doc comment should be the same on both members.

I'll be opening a PR against azure/azure-sdk for this as well.

**Comment by KrzysztofCwalina** [comment]
We should also have a higher bar for using these extensible enums. They cause lots of problems in UX.

**Comment by heaths** [comment]
As discussed internally, we'll use `Values` instead.

As for a higher bar, what should we use instead? We already state they have to be used for input or roundtripping. .NET without unions gives us no other options unless we want to overload parameters (enum or string), and use only strings in model properties, but then how would people know they can use enum values (via `ToString()`)?

**Comment by heaths** [comment]
With https://github.com/Azure/azure-sdk/pull/1610 merged, this is now part of the guidelines: <https://azure.github.io/azure-sdk/dotnet_implementation.html#dotnet-enums-values>

Having an attribute or something as opposed to plumbing through an AutoRest directive would be fine, especially since this may only affect C# anyway.

**Comment by pakrym** [comment]
Make this conditional on `Values` type defined under the extensible enum.

**Comment by heaths** [comment]
How would that work, though? Can the nested class also be partial?

**Comment by pakrym** [comment]
Yup, https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BYAUAAQGYACfARgDZiAHCBGASwgBtSAmUsgdmIG89ig0iXJVa9Jq3wcACggD2AcwQQwHXgF88WgiMo06jFu049+uIcM5jDkk3KUq1fHRqA=

**Comment by heaths** [comment]
Cool (I keep forgetting about that site until someone links it like that 🤦‍♂️). Sounds like a good idea, and no additional attributes necessary. :shipit:
